### PR TITLE
Refactor: Trajectory loading seperated into external function

### PR DESCRIPTION
**Pull request by onurtore**
Utilizes additional function to load actor trajectories.

### CODE REVIEW

File: include/ignition/gazebo/rendering/SceneManager.hh [hunk 1]
```diff
@@ -23,6 +23,7 @@
 #include <string>
 #include <utility>
 #include <vector>
+#include <unordered_map>
```
Comment (ahcorde) [inline include/ignition/gazebo/rendering/SceneManager.hh:26]: alphabetize

File: src/rendering/SceneManager.cc [hunk 2]
```diff
@@ -2336,3 +2253,102 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
   animData.valid = true;
   return animData;
 }
+
+std::vector<common::TrajectoryInfo>
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2257]:
```suggestion
/////////////////////////////////////////////////
std::vector<common::TrajectoryInfo>
```

File: src/rendering/SceneManager.cc [hunk 3]
```diff
@@ -2336,3 +2253,102 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+SceneManager::LoadTrajectories(const sdf::Actor &_actor,
+  std::unordered_map<std::string, unsigned int> &_mapAnimNameId,
+  Entity &_id)
+{
+
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2262]:
```suggestion
```

File: src/rendering/SceneManager.cc [hunk 4]
```diff
@@ -2336,3 +2253,102 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+  common::MeshManager *meshManager = common::MeshManager::Instance();
+  descriptor.mesh = meshManager->Load(descriptor.meshName);
+  common::SkeletonPtr meshSkel = descriptor.mesh->MeshSkeleton();
+
+  std::vector<common::TrajectoryInfo> trajectories;
+
+  if (_actor.TrajectoryCount() != 0)
+  {
+    // Load all trajectories specified in sdf
+    for (unsigned i = 0; i < _actor.TrajectoryCount(); i++)
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2274]:
```suggestion
    for (unsigned i = 0; i < _actor.TrajectoryCount(); ++i)
```

File: src/rendering/SceneManager.cc [hunk 5]
```diff
@@ -2336,3 +2253,102 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+      }
+
+      common::TrajectoryInfo trajInfo;
+      trajInfo.SetId(trajSdf->Id());
+      trajInfo.SetAnimIndex(_mapAnimNameId[trajSdf->Type()]);
+
+      if (trajSdf->WaypointCount() != 0)
+      {
+        std::map<TP, math::Pose3d> waypoints;
+        for (unsigned j = 0; j < trajSdf->WaypointCount(); j++)
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2291]:
```suggestion
        for (unsigned j = 0; j < trajSdf->WaypointCount(); ++j)
```

File: src/rendering/SceneManager.cc [hunk 6]
```diff
@@ -2336,3 +2253,102 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+        {
+          auto point = trajSdf->WaypointByIndex(j);
+          TP pointTp(std::chrono::milliseconds(
+                    static_cast<int>(point->Time()*1000)));
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2295]:
```suggestion
                    static_cast<int>(point->Time() * 1000)));
```

File: src/rendering/SceneManager.cc [hunk 7]
```diff
@@ -2336,3 +2253,102 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+          waypoints[pointTp] = point->Pose();
+        }
+        trajInfo.SetWaypoints(waypoints, trajSdf->Tension());
+        // Animations are offset by 1 because index 0 is taken by the mesh name
+        auto animation = _actor.AnimationByIndex(trajInfo.AnimIndex()-1);
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2300]:
```suggestion
        auto animation = _actor.AnimationByIndex(trajInfo.AnimIndex() - 1);
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: src/rendering/SceneManager.cc (hunk 3 above)
Comment (onurtore) [reply src/rendering/SceneManager.cc]: Done.

File: src/rendering/SceneManager.cc (hunk 4 above)
Comment (onurtore) [reply src/rendering/SceneManager.cc:2274]: Is there a specific reason for that? Is it safer?

Comment (onurtore) [reply src/rendering/SceneManager.cc]: Done.

File: src/rendering/SceneManager.cc (hunk 5 above)
Comment (onurtore) [reply src/rendering/SceneManager.cc]: Done.

File: src/rendering/SceneManager.cc (hunk 6 above)
Comment (onurtore) [reply src/rendering/SceneManager.cc:2295]: Done.

File: src/rendering/SceneManager.cc (hunk 7 above)
Comment (onurtore) [reply src/rendering/SceneManager.cc]: Done.

**Review by ahcorde — CHANGES_REQUESTED** [review]
conflicts

**Comment by onurtore** [comment]
Commits done by Github Interface fails to comply DCO. Should I update them?

File: src/rendering/SceneManager.cc [hunk 8]
```diff
@@ -2243,7 +2260,6 @@ SceneManager::LoadAnimations(const sdf::Actor &_actor)
   common::MeshManager *meshManager = common::MeshManager::Instance();
   descriptor.mesh = meshManager->Load(descriptor.meshName);
   common::SkeletonPtr meshSkel = descriptor.mesh->MeshSkeleton();
-
```
Comment (ahcorde) [inline src/rendering/SceneManager.cc:2246]: This change is not related with the PR, do you mind to remove it ?

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: src/rendering/SceneManager.cc (hunk 8 above)
Comment (onurtore) [reply src/rendering/SceneManager.cc]: Done.

**Comment by onurtore** [comment]
@ahcorde  Friendly ping

File: include/ignition/gazebo/rendering/SceneManager.hh [hunk 9]
```diff
@@ -214,6 +214,14 @@ inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
         std::vector<common::TrajectoryInfo>& _trajectories,
         std::chrono::steady_clock::time_point _time);
 
+    /// \brief Load Actor trajectories
+    /// \param[in] _actor Actor
+    /// \return Trajectory vector
+    public: std::vector<common::TrajectoryInfo>
+        LoadTrajectories(const sdf::Actor &_actor,
+        std::unordered_map<std::string, unsigned int> &_mapAnimNameId,
+        Entity &_id);
```
Comment (chapulina) [inline include/ignition/gazebo/rendering/SceneManager.hh:223]: Can this be a private function inside `SceneManagerPrivate`? This header is installed, so whatever we add here will be hard to change in the future. But whatever is in the `.cc` file can be changed at any time!

File: include/ignition/gazebo/rendering/SceneManager.hh [hunk 10]
```diff
@@ -214,6 +214,14 @@ inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
         std::vector<common::TrajectoryInfo>& _trajectories,
         std::chrono::steady_clock::time_point _time);
 
+    /// \brief Load Actor trajectories
+    /// \param[in] _actor Actor
```
Comment (chapulina) [inline include/ignition/gazebo/rendering/SceneManager.hh:218]: Can you document all params? Also I believe `_id` doesn't need to be passed as a reference.

File: src/rendering/SceneManager.cc [hunk 11]
```diff
@@ -2235,6 +2153,105 @@ AnimationUpdateData SceneManagerPrivate::ActorTrajectoryAt(
...
+std::vector<common::TrajectoryInfo>
+SceneManager::LoadTrajectories(const sdf::Actor &_actor,
+  std::unordered_map<std::string, unsigned int> &_mapAnimNameId,
+  Entity &_id)
+{
+  rendering::MeshDescriptor descriptor;
+  descriptor.meshName = asFullPath(_actor.SkinFilename(), _actor.FilePath());
+  common::MeshManager *meshManager = common::MeshManager::Instance();
+  descriptor.mesh = meshManager->Load(descriptor.meshName);
+  common::SkeletonPtr meshSkel = descriptor.mesh->MeshSkeleton();
```
Comment (chapulina) [inline src/rendering/SceneManager.cc:2166]: The logic to get a skeleton from a descriptor is already done outside the function at line 963. Can't we just pass the skeleton got there to the function, so we don't need to retrieve it again?

**Review by chapulina — CHANGES_REQUESTED** [review]

File: src/rendering/SceneManager.cc (hunk 11 above)
Comment (onurtore) [reply src/rendering/SceneManager.cc]: Done.

File: include/ignition/gazebo/rendering/SceneManager.hh (hunk 10 above)
Comment (onurtore) [reply include/ignition/gazebo/rendering/SceneManager.hh]: Done.

File: include/ignition/gazebo/rendering/SceneManager.hh (hunk 9 above)
Comment (onurtore) [reply include/ignition/gazebo/rendering/SceneManager.hh]: Done.

**Comment by onurtore** [comment]
DCO failing. I tried to rebase but ended up with many conflicts which are not related to this PR.

**Comment by ahcorde** [comment]
@osrf-jenkins run tests please!

**Review by ahcorde — APPROVED** [review]
